### PR TITLE
Enable Hash Validation in UCM

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

Enables entity hash validation in UCM by default, previously it simply trusted Share.

This provides us as Share and UCM implementers with an additional protection such that corrupted hashes from UCM or Share bugs won't propagate through the community and corrupt codebases. It also ensures we'll learn about such bugs as soon as possible.

Performance impact is minimal since generally the hash validation can take place while we would otherwise be blocked on IO/network.

E.g. here's me pulling base/main from share with and without hash validation:

```
Clone base/main
    with hash verification:
        1m8.32s real            54.62s user             10.68s sys
    without hash verification:
        1m4.67s real            34.57s user             8.87s sys
```

We see that the `user` work is higher, but the realtime is only marginally affected. On many subsequent runs they would switch places, sometimes the validated UCM performed better than the unvalidated one.

Here's an example of what a hash-validation error looks like (I just cloned base without the map of known mismatches):

```
.> clone @unison/base

  ❗️

  The hash associated with the given term component entity is incorrect.

  The associated hash is:
  #4h9sgktb6jb2u9ubblo5pj3jl9nkjvr2coa8u0fkt0hjl8v1i26jppbu190fmrt8tbts3dfmqm4ohdeeq52k5hv93rt6auvaacsniig
  The computed hash is:
  #243evnehcbtmk52049o5938vijh89pi75cvu82nev5esi42c4vokvjj7fgj29tc77nmpupktmomht34to2a4i498mcdju3f753gi6mg

  Please create an issue and report this to the Unison team.
```

## Implementation notes

* Include a module with a hard-coded map of all known component mismatches from the bugged UCM. This list hasn't changed in a few months, so we probably don't need it to be dynamically updatable, at least not for now.
* Flips the default of `UNISON_ENTITY_VALIDATION` env var, now you need to `UNISON_ENTITY_VALIDATION=false` to DISABLE it.
* Fixes UCM hash validation, previously it was only validating the first causal hash, now it validates every entity.

## Interesting/controversial decisions

Including the list of known mismatches could be controversial I suppose, but it's one way to provide a reasonable UX in the meantime until we can roll out a rehash migration.

## Test coverage

I've cloned `@unison/base/main`, `@unison/cloud/main` and `@unison/distributed/main` without any issues. 

## Loose ends

Link to related issues that address things you didn't get to. Stuff you encountered on the way and decided not to include in this PR.

### CODE REVIEW

File: unison-share-api/src/Unison/Sync/EntityValidation.hs [hunk 1]
```diff
@@ -70,7 +70,7 @@ validatePatchFull expectedHash32 localIds bytes = do
             HH.hashPatchFormatFull v2HashHandle localIds' localPatch
       if actualHash == PatchHash expectedHash
         then Nothing
-        else Just $ Share.EntityHashMismatch Share.NamespaceType (mismatch expectedHash (unPatchHash actualHash))
+        else Just $ Share.EntityHashMismatch Share.PatchType (mismatch expectedHash (unPatchHash actualHash))
```
Comment (ChrisPenner) [inline unison-share-api/src/Unison/Sync/EntityValidation.hs:73]: These were just resulting in incorrect error messages.

File: stack.yaml [hunk 2]
```diff
@@ -70,3 +70,5 @@ ghc-options:
 
  # See https://github.com/haskell/haskell-language-server/issues/208
  "$everything": -haddock
+
+ statistics: -fsimpl-tick-factor=10000 # statistics fails on GHC 9 without this, https://github.com/haskell/statistics/issues/173
```
Comment (ChrisPenner) [inline stack.yaml:74]: Known issue with the statistics package that makes it fail to build on `stack build --profile` runs, this fixes it, can probably remove this in a future snapshot if they sort it out on the package side.

File: docs/configuration.md [hunk 3]
```diff
@@ -108,10 +108,11 @@ $ UNISON_READONLY="true" ucm
 
 ### `UNISON_ENTITY_VALIDATION`
 
-Enable validation of entities pulled from a codebase server.
+Allows disabling validation of entities pulled from a codebase server.
+Defaults to enabled.
```
Comment (aryairani) [inline docs/configuration.md]: "It's generally not a good idea to disable this."

File: unison-cli/src/Unison/Share/ExpectedHashMismatches.hs
Comment (aryairani) [inline unison-cli/src/Unison/Share/ExpectedHashMismatches.hs:1]: Wow there's so many!

File: unison-cli/src/Unison/Share/Sync.hs [hunk 4]
```diff
@@ -74,13 +75,13 @@ import Unison.Util.Monoid (foldMapM)
...
 maxSimultaneousPullDownloaders :: Int
-maxSimultaneousPullDownloaders = 5
+maxSimultaneousPullDownloaders = 1
 
 -- | The maximum number of push workers at a time. Each push worker reads from the database and uploads entities.
 -- Share currently parallelizes on it's own in the backend, and any more than one push worker
 -- just results in serialization conflicts which slow things down.
 maxSimultaneousPushWorkers :: Int
-maxSimultaneousPushWorkers = 5
+maxSimultaneousPushWorkers = 1
```
Comment (aryairani) [inline unison-cli/src/Unison/Share/Sync.hs]: Did we wanna leave these at 1?

File: unison-cli/src/Unison/Share/Sync.hs [hunk 5]
```diff
@@ -624,10 +631,13 @@ completeTempEntities httpClient unisonShareUrl connect repoInfo downloadedCallba
           pure (Left (SyncError err))
         Right (Share.DownloadEntitiesSuccess entities) -> do
           downloadedCallback (NESet.size hashes)
-          atomically do
-            writeTQueue entitiesQueue (hashes, entities)
-            recordNotWorking workerCount
-          pure (Right ())
+          case validateEntities entities of
+            Left err -> pure . Left . SyncError . Share.DownloadEntitiesEntityValidationFailure $ err
```
Comment (aryairani) [inline unison-cli/src/Unison/Share/Sync.hs:640]: Just out of curiosity, what does this error look like?

**Review by aryairani — APPROVED** [review]

File: unison-cli/src/Unison/Share/Sync.hs (hunk 4 above)
Comment (ChrisPenner) [reply unison-cli/src/Unison/Share/Sync.hs]: Oh whoops, we want to leave `pull` at 5, but push goes to 1 👍🏼

File: unison-cli/src/Unison/Share/Sync.hs (hunk 5 above)
Comment (ChrisPenner) [reply unison-cli/src/Unison/Share/Sync.hs:640]: Good question!  Here's me cloning base with an empty map of known mismatches:



```
.> clone @unison/base

  ❗️

  The hash associated with the given term component entity is incorrect.

  The associated hash is:
  #4h9sgktb6jb2u9ubblo5pj3jl9nkjvr2coa8u0fkt0hjl8v1i26jppbu190fmrt8tbts3dfmqm4ohdeeq52k5hv93rt6auvaacsniig
  The computed hash is:
  #243evnehcbtmk52049o5938vijh89pi75cvu82nev5esi42c4vokvjj7fgj29tc77nmpupktmomht34to2a4i498mcdju3f753gi6mg

  Please create an issue and report this to the Unison team.
```

**Pull request merged**